### PR TITLE
[BEAM-2966] Allow subclasses of tuple, list, and dict as pvaluish inputs/outputs.

### PR DESCRIPTION
**Pull request by robertwb**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @KesterTong

**Comment by robertwb** [comment]
jenkins: retest this please

**Comment by robertwb** [comment]
jenkins: retest this please

File: sdks/python/apache_beam/transforms/ptransform.py [hunk 1]
```diff
@@ -73,27 +73,25 @@ class _PValueishTransform(object):
...
-    return [self.visit(x, *args) for x in node]
-
-  def visit_tuple(self, node, *args):
-    return tuple(self.visit(x, *args) for x in node)
-
-  def visit_dict(self, node, *args):
-    return {key: self.visit(value, *args) for (key, value) in node.items()}
+  def visit_nested(self, node, *args):
+    if isinstance(node, (tuple, list)):
+      # namedtuples require unpacked arguments in their constructor,
```
Comment (KesterTong) [inline sdks/python/apache_beam/transforms/ptransform.py]: It's not clear that this supports subclasses of a namedtuple.  Such subclasses will inherit the _make class method from the namedtuple, but the _make method will produce the namedtuple not the subclass.  Instead, could we test for the existence of _make (as a test of whether we are dealing with a subclass of namedtuple or a direct subclass of tuple, which we assume has the usual tuple constructor) but still invoke the constructor `node.__class__` when dealing with a subclass of a namedtuple? i.e.

```
if isinstance(node, tuple) and hasattr(node.__class__, '_make'):
  # node is an instance of a subclass of a namedtuple.
  return node.__class__(*[self.visit(x, *args) for x in node])
elif isinstance(node, (tuple, list)):
  ...
```

**Comment by KesterTong** [comment]
is there a JIRA issue for this PR?  If not I could open one.  Would be helpful as reference for tf.Transform release notes etc. since tf.Transform will rely on older versions of Beam which will hit the bug this PR fixes.

File: sdks/python/apache_beam/transforms/ptransform.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/ptransform.py]: Done.

**Comment by robertwb** [comment]
PTAL

**Comment by KesterTong** [comment]
Thanks, I think you missed my question above regarding the commit titled "Support multiple materializations of the smae pvalue."?

**Comment by robertwb** [comment]
Sorry, I added comments but did not address it directly. The cache is solely an implementation detail that is thrown away as soon as the pipeline is gc'd (and, hopefully, will simply go away completely when we clean things up). In particular, the ref-counting is only used during pipeline execution.

**Comment by robertwb** [comment]
retest this please

**Pull request closed without merging**